In [90]:
#import library and data
import pandas as pd
import numpy as np

product_data = pd.read_csv(
    "C:\\Users\\Admin\\Desktop\\DA books\\DA lesson\\Python\\product.csv"
)
transactions_data = pd.read_csv(
    "C:\\Users\\Admin\\Desktop\\DA books\\DA lesson\\Python\\transactions.csv"
)
payment_data = pd.read_csv(
    "C:\\Users\\Admin\\Desktop\\DA books\\DA lesson\\Python\\payment_report.csv"
)

In [91]:
#Check product data

print(product_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 492 entries, 0 to 491
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   product_id  492 non-null    int64 
 1   category    492 non-null    object
 2   team_own    492 non-null    object
dtypes: int64(1), object(2)
memory usage: 11.7+ KB
None


In [92]:
#check payment_report data
print(payment_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 919 entries, 0 to 918
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   report_month   919 non-null    object
 1   payment_group  919 non-null    object
 2   product_id     919 non-null    int64 
 3   source_id      919 non-null    int64 
 4   volume         919 non-null    int64 
dtypes: int64(3), object(2)
memory usage: 36.0+ KB
None


In [93]:
#check transactions data
print(transactions_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1324002 entries, 0 to 1324001
Data columns (total 9 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   transaction_id  1324002 non-null  int64  
 1   merchant_id     1324002 non-null  int64  
 2   volume          1324002 non-null  int64  
 3   transType       1324002 non-null  int64  
 4   transStatus     1324002 non-null  int64  
 5   sender_id       1274943 non-null  float64
 6   receiver_id     1159207 non-null  float64
 7   extra_info      6095 non-null     object 
 8   timeStamp       1324002 non-null  int64  
dtypes: float64(2), int64(6), object(1)
memory usage: 90.9+ MB
None


In [94]:
#Join product.csv and payment_report.csv, explore data  
payment_enriched = pd.merge(payment_data, product_data, on = "product_id")
payment_enriched["report_month"] = pd.to_datetime(
    payment_enriched["report_month"]
)

In [95]:
#Payment enriched data
#check missing values, datatype
print(payment_enriched.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 897 entries, 0 to 896
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   report_month   897 non-null    datetime64[ns]
 1   payment_group  897 non-null    object        
 2   product_id     897 non-null    int64         
 3   source_id      897 non-null    int64         
 4   volume         897 non-null    int64         
 5   category       897 non-null    object        
 6   team_own       897 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(3)
memory usage: 49.2+ KB
None


In [96]:
#check incorrect values
print(payment_enriched.describe())

                     report_month    product_id  source_id        volume
count                         897    897.000000      897.0  8.970000e+02
mean   2023-02-19 07:45:33.110368   1139.573021       45.0  1.338153e+08
min           2023-01-01 00:00:00     12.000000       45.0  1.000000e+04
25%           2023-02-01 00:00:00    634.000000       45.0  1.258000e+06
50%           2023-03-01 00:00:00   1023.000000       45.0  7.469786e+06
75%           2023-04-01 00:00:00   1578.000000       45.0  4.770741e+07
max           2023-04-01 00:00:00  15067.000000       45.0  4.926051e+09
std                           NaN   1161.547355        0.0  4.614672e+08


In [97]:
#change data type
payment_enriched["payment_group"] = payment_enriched["payment_group"].astype('string')
payment_enriched["category"] = payment_enriched["category"].astype('string')
payment_enriched["team_own"] = payment_enriched["team_own"].astype('string')

In [98]:
#Recheck duplicate
print("Duplicate data is: " + str(payment_enriched.duplicated().sum()))

Duplicate data is: 0
Null data in:
report_month     0
payment_group    0
product_id       0
source_id        0
volume           0
category         0
team_own         0
dtype: int64


In [ ]:
#Recheck missing values
print("Null data in:")
print(payment_enriched.isna().sum())

In [116]:
#Recheck datatype

pd.set_option('display.max_columns', None)  
pd.set_option('display.width', 1000)

print(payment_enriched.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 897 entries, 0 to 896
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   report_month   897 non-null    datetime64[ns]
 1   payment_group  897 non-null    string        
 2   product_id     897 non-null    int64         
 3   source_id      897 non-null    int64         
 4   volume         897 non-null    int64         
 5   category       897 non-null    string        
 6   team_own       897 non-null    string        
dtypes: datetime64[ns](1), int64(3), string(3)
memory usage: 49.2 KB
None


In [115]:
#Recheck incorrect values
print(payment_enriched.describe())

                     report_month    product_id  source_id        volume
count                         897    897.000000      897.0  8.970000e+02
mean   2023-02-19 07:45:33.110368   1139.573021       45.0  1.338153e+08
min           2023-01-01 00:00:00     12.000000       45.0  1.000000e+04
25%           2023-02-01 00:00:00    634.000000       45.0  1.258000e+06
50%           2023-03-01 00:00:00   1023.000000       45.0  7.469786e+06
75%           2023-04-01 00:00:00   1578.000000       45.0  4.770741e+07
max           2023-04-01 00:00:00  15067.000000       45.0  4.926051e+09
std                           NaN   1161.547355        0.0  4.614672e+08


<!-- Comment for payment_enriched data -->
Regarding to payment_enriched data (Joining 2 table):
Missing values : 0 values   => No action
Duplicates : 0 values => No action
Incorect data types : 0 values=> No action
Incorrect Values : 0 values => No action

In [100]:
#Transactions data
#check missing values, datatype
print(transactions_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1324002 entries, 0 to 1324001
Data columns (total 9 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   transaction_id  1324002 non-null  int64  
 1   merchant_id     1324002 non-null  int64  
 2   volume          1324002 non-null  int64  
 3   transType       1324002 non-null  int64  
 4   transStatus     1324002 non-null  int64  
 5   sender_id       1274943 non-null  float64
 6   receiver_id     1159207 non-null  float64
 7   extra_info      6095 non-null     object 
 8   timeStamp       1324002 non-null  int64  
dtypes: float64(2), int64(6), object(1)
memory usage: 90.9+ MB
None


In [101]:

#Check incorrect values, 
print(transactions_data.describe())

       transaction_id   merchant_id        volume     transType   transStatus     sender_id   receiver_id     timeStamp
count    1.324002e+06  1.324002e+06  1.324002e+06  1.324002e+06  1.324002e+06  1.274943e+06  1.159207e+06  1.324002e+06
mean     3.002233e+09  2.460318e+03  2.388059e+05  6.979222e+00 -1.204625e+01  1.033938e+08  2.084884e+08  1.683130e+12
std      1.042606e+07  3.304277e+03  9.681009e+05  7.459714e+00  5.577823e+01  6.234305e+08  9.287794e+08  1.707815e+08
min      3.000000e+09  5.000000e+00  1.000000e+00  2.000000e+00 -1.333000e+03  1.000001e+07 -6.300000e+01  1.682874e+12
25%      3.001121e+09  3.050000e+02  1.000000e+04  2.000000e+00  1.000000e+00  1.005657e+07  1.526700e+05  1.682994e+12
50%      3.002200e+09  2.250000e+03  3.000000e+04  2.000000e+00  1.000000e+00  1.094193e+07  7.025130e+06  1.683097e+12
75%      3.003255e+09  2.270000e+03  1.000000e+05  8.000000e+00  1.000000e+00  2.101266e+07  3.151238e+07  1.683269e+12
max      6.000066e+09  1.625250e+05  7.8

In [102]:
#find duplicates data
print(transactions_data.isna().sum())

print("Num of Duplicated data is: " + str(transactions_data.duplicated().sum()))

transaction_id          0
merchant_id             0
volume                  0
transType               0
transStatus             0
sender_id           49059
receiver_id        164795
extra_info        1317907
timeStamp               0
dtype: int64
Num of Duplicated data is: 28


In [103]:

#change data type, fill null, convert incorrect values(negative to positive)
transactions_data["receiver_id"] = transactions_data["receiver_id"].abs()
transactions_data["transStatus"] = transactions_data["transStatus"].abs()
transactions_data["sender_id"] = transactions_data["sender_id"].fillna(0).astype(int)
transactions_data["receiver_id"] = transactions_data["receiver_id"].fillna(0).astype(int)

In [104]:
#remove dup
transaction_remove_dup = transactions_data.drop_duplicates(subset= "transaction_id")

In [ ]:
#Recheck duplicates

pd.set_option('display.max_columns', None)  
pd.set_option('display.width', 1000)

print("Num of Removed duplicated data is :" + str(transaction_remove_dup.duplicated().sum()))

Num of Removed duplicated data is :0


In [106]:
#Recheck missing values , datatype
print("Null data in:")
print(transaction_remove_dup.info())

print(transaction_remove_dup.isna().sum())

Null data in:
<class 'pandas.core.frame.DataFrame'>
Index: 1323974 entries, 0 to 1324001
Data columns (total 9 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   transaction_id  1323974 non-null  int64 
 1   merchant_id     1323974 non-null  int64 
 2   volume          1323974 non-null  int64 
 3   transType       1323974 non-null  int64 
 4   transStatus     1323974 non-null  int64 
 5   sender_id       1323974 non-null  int64 
 6   receiver_id     1323974 non-null  int64 
 7   extra_info      6095 non-null     object
 8   timeStamp       1323974 non-null  int64 
dtypes: int64(8), object(1)
memory usage: 101.0+ MB
None
transaction_id          0
merchant_id             0
volume                  0
transType               0
transStatus             0
sender_id               0
receiver_id             0
extra_info        1317879
timeStamp               0
dtype: int64


In [107]:
#Recheck incorrect values
print(transaction_remove_dup.describe())

       transaction_id   merchant_id        volume     transType   transStatus     sender_id   receiver_id     timeStamp
count    1.323974e+06  1.323974e+06  1.323974e+06  1.323974e+06  1.323974e+06  1.323974e+06  1.323974e+06  1.323974e+06
mean     3.002234e+09  2.460332e+03  2.388093e+05  6.979182e+00  1.393273e+01  9.956116e+07  1.825382e+08  1.683130e+12
std      1.042617e+07  3.304293e+03  9.681107e+05  7.459665e+00  5.533747e+01  6.120781e+08  8.717817e+08  1.707816e+08
min      3.000000e+09  5.000000e+00  1.000000e+00  2.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00  1.682874e+12
25%      3.001121e+09  3.050000e+02  1.000000e+04  2.000000e+00  1.000000e+00  1.004014e+07  4.059475e+04  1.682994e+12
50%      3.002200e+09  2.250000e+03  3.000000e+04  2.000000e+00  1.000000e+00  1.057187e+07  3.529736e+06  1.683097e+12
75%      3.003255e+09  2.270000e+03  1.000000e+05  8.000000e+00  1.000000e+00  2.100163e+07  2.451345e+07  1.683269e+12
max      6.000066e+09  1.625250e+05  7.8

<!-- Comment for transactions_data -->
Regarding to transacion_data :
Missing data : sender_id (49059). receiver_id (164795), extra_info (1317907)  => fill null sender_id, receiver_id
Duplicates : 28 values => Remove duplicates
Incorect data types : sender_id. receiver_id (Float) => Convert to (Int) data type
Incorrect Values : receiver_id, transStatus gets negative values  => Convert to suitable values (positive values)


In [108]:
#sum volume and get top 3 products
total_volume = pd.DataFrame(payment_data.groupby("product_id")["volume"].sum())
top_3_productids = total_volume.sort_values("volume", ascending= False).head(3)

print(top_3_productids)


                 volume
product_id             
1976        61797583647
429         14667676567
372         13713658515


In [109]:
#total unique data, product owned by 1 team
total_unique = payment_enriched.groupby(
    "product_id", as_index = False)["team_own"].nunique()
team_normal = total_unique[total_unique["team_own"] == 1]
team_abnormal = total_unique[total_unique["team_own"] != 1]

pd.set_option('display.max_columns', None)  
pd.set_option('display.width', 1000)

print(total_unique)
print(team_normal)
print(team_abnormal)

     product_id  team_own
0            12         1
1            15         1
2            17         1
3            18         1
4            19         1
..          ...       ...
303        2408         1
304        2419         1
305        2587         1
306       10039         1
307       15067         1

[308 rows x 2 columns]
     product_id  team_own
0            12         1
1            15         1
2            17         1
3            18         1
4            19         1
..          ...       ...
303        2408         1
304        2419         1
305        2587         1
306       10039         1
307       15067         1

[308 rows x 2 columns]
Empty DataFrame
Columns: [product_id, team_own]
Index: []


In [110]:
#total volume by team since 2023-04
total = pd.DataFrame(
    payment_enriched[payment_enriched["report_month"].dt.quarter >= 2]
    .groupby("team_own")["volume"]
    .sum()
)
total["rank"] = total.sort_values("volume").rank()
print(total[total["rank"]== 1])

            volume  rank
team_own                
APS       51141753   1.0


In [111]:
#find category have lowest contribution to that team (APS)
total1 = pd.DataFrame(
    payment_enriched[
        (payment_enriched["report_month"].dt.quarter >= 2) 
        & (payment_enriched["team_own"] == "APS")
    ]
    .groupby("category")["volume"]
    .sum()
)
total1["rank1"] = total1.sort_values("volume").rank()
print(total1[total1["rank1"]== 1])

            volume  rank1
category                 
PXXXXXE   25232438    1.0


In [112]:
contribution = pd.DataFrame(
    payment_data[payment_data["payment_group"] == "refund"]
    .groupby("source_id")["volume"]
    .sum()
)
contribution["rank_cont"] = contribution.sort_values("volume", ascending = False).rank(ascending= False)
print(contribution[contribution["rank_cont"] == 1])

                volume  rank_cont
source_id                        
38         36527454759        1.0


In [113]:
#list of categories
transactions_data["sender_id"] = transactions_data["sender_id"].fillna(0).astype(int)
transactions_data["receiver_id"] = transactions_data["receiver_id"].fillna(0).astype(int)
transaction_remove_dup = transactions_data.drop_duplicates(subset= "transaction_id")
transaction_categories = [
    "Bank Transfer Transaction", 
    "Withdraw Money Transaction", 
    "Top Up Money Transaction",
    "Payment Transaction",
    "Transfer Money Transaction", 
    "Split Bill Transaction"
]
#list of conditions, through by each position in list
conditions = [
    (transaction_remove_dup["transType"] == 2) 
    & (transaction_remove_dup["merchant_id"] == 1205),
    (transaction_remove_dup["transType"] == 2) 
    & (transaction_remove_dup["merchant_id"] == 2260),
    (transaction_remove_dup["transType"] == 2) 
    & (transaction_remove_dup["merchant_id"] == 2270),
    (transaction_remove_dup["transType"] == 2) 
    & (~transaction_remove_dup["merchant_id"].isin([1205, 2260, 2270])),
    (transaction_remove_dup["transType"] == 8) 
    & (transaction_remove_dup["merchant_id"] == 2250),
    (transaction_remove_dup["transType"] == 8) 
    & (~transaction_remove_dup["merchant_id"].isin([2250]))]
transaction_remove_dup["transaction_type"]= np.select(
    conditions, transaction_categories, default = "Invalid Transactions"
)

pd.set_option('display.max_columns', None)  
pd.set_option('display.width', 1000)

print(transaction_remove_dup)

         transaction_id  merchant_id   volume  transType  transStatus  sender_id  receiver_id extra_info      timeStamp            transaction_type
0            3002692434            5   100000         24            1   10199794       199794        NaN  1682932054455        Invalid Transactions
1            3002692437          305    20000          2            1   14022211     14022211        NaN  1682932054912         Payment Transaction
2            3001960110         7255    48605         22            1          0     10530940        NaN  1682932055000        Invalid Transactions
3            3002680710         2270  1500000          2            1   10059206        59206        NaN  1682932055622    Top Up Money Transaction
4            3002680713         2275    90000          2            1   10004711         4711        NaN  1682932056197         Payment Transaction
...                 ...          ...      ...        ...          ...        ...          ...        ...        

C:\Users\Admin\AppData\Local\Temp\ipykernel_9120\2163614259.py:27: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  transaction_remove_dup["transaction_type"]= np.select(


In [114]:
#num of transactions, volume, senders, receivers unique.
summarize_transaction = (
    transaction_remove_dup[
        transaction_remove_dup["transaction_type"] != "Invalid Transactions"
    ]
    .groupby("transaction_type")
    .agg(
        num_transactions = ("transaction_id", "nunique"),
        total_volume = ("volume", "sum"),
        num_senders = ("sender_id", "nunique"),
        num_receivers = ("receiver_id", "nunique")
    )
)

pd.set_option('display.max_columns', None)  
pd.set_option('display.width', 1000)

print(summarize_transaction)

                            num_transactions  total_volume  num_senders  num_receivers
transaction_type                                                                      
Bank Transfer Transaction              37879   50605806190        23156           9272
Payment Transaction                   398665   71850608441       139583         113297
Split Bill Transaction                  1376       4901464         1323            572
Top Up Money Transaction              290498  108605618829       110409         110409
Transfer Money Transaction            341173   37032880492        39021          34585
Withdraw Money Transaction             33725   23418181420        24814          24814


# Conclusion
# EDA Results
# - Regarding to payment_enriched data (Joining 2 payment_report.csv and product.csv file), it can be seen that no missing values. Moreover, there were no duplicates, inccorect data type as well as incorrect values in these dataset, so no actions needed for these exploratory.
# - According to transacion_data : missing data is identified in sender_id (49059). receiver_id (164795), extra_info (1317907) columns, however just sender_id and receiver_id are needed for filling null values action. Besides, the dataset contained 28 duplicates rows, so that is required to be removed for enhacing data accuracy. Furthermore, sender_id and receiver_id need to convert from float to Integer data type for clear meaning. Especially, negative values are found in receiver_id, transStatus , these are inccorected values that need to be converted to positive values.
# #Key Finding
#  - Top three products ID: with the highest transaction volume are 1976, 429 and 372 orderly. From these consequence , Company can prioritize product marketing, marketing campaigns for next seasons as well as resource allocation.
#  - Performance: since Q2 2023,  APS team can be seen as the worst performance with the lowest volume, the least productive category is PXXXXXE. This team should attention on the next strategies to have a better improvement, reach target to enhance team performance
#  - Refund Transactions analysis: Moreover , The analysis of refund transactions indicated that the source ID contributing the most was 38, with the volume of 36,527,454,759. Understanding the sources and reasons behind these transactions can inform strategies to minimize future occurrences, reduce operational costs, and enhance customer satisfaction.
#  - Payment , Top up Money and Transfer Money are the three main reasons customers use e-wallet, within Payment have the highest number of transactions, while most people Top Up Money the the greatest amount of money to e-wallet for various daily expense.
#  - Overall, These findings emphasize key factors relating to product performance, team performance, contribution and transactions dynamic within e-wallet service. By seeking for, analysing and addressing many identified issues and difficult challenges as well as utilizing the insights gained, the company can improve efficiency, team performance in order to devote overall financial performance as well as company development.
 
 